pkg/restore/restore.go

```diff
@@ -578,6 +590,94 @@ func addToResult(r *api.RestoreResult, ns string, e error) {
 	}
 }
 
+func (ctx *context) waitForPVDeletion(name string, resourceClient client.Dynamic) (bool, error) {
+	pvLogger := ctx.log.WithField("pvName", name)
+
+	var shouldRestore bool
+	err := wait.PollImmediate(time.Second, ctx.resourceTerminatingTimeout, func() (bool, error) {
+		clusterPV, err := resourceClient.Get(name, metav1.GetOptions{})
+
+		if apierrors.IsNotFound(err) {
+			pvLogger.Debug("PV not found, safe to restore")
+			// PV not found, can safely exit loop and proceed with restore.
+			shouldRestore = true
+			return true, nil
+		}
+
+		if err != nil {
+			return false, errors.Wrapf(err, "could not retrieve in-cluster copy of PV %s", name)
+
+		}
+
+		phase, err := collections.GetString(clusterPV.UnstructuredContent(), "status.phase")
+		if err != nil {
+			// Break the loop since we couldn't read the phase
+			return true, errors.Wrapf(err, "error getting phase for in-cluster PV %s", name)
+		}
+
+		if phase == string(v1.VolumeReleased) || clusterPV.GetDeletionTimestamp() != nil {
+			// PV was found and marked for deletion, or it was released; wait for it to go away.
+			pvLogger.Debugf("PV found, but marked for deletion, waiting")
+			return false, nil
+		}
+
+		namespace, err := collections.GetString(clusterPV.UnstructuredContent(), "spec.claimRef.namespace")
+		if err != nil {
+			return false, errors.Wrapf(err, "error looking up namespace name for in-cluster PV %s", name)
+		}
+		pvcName, err := collections.GetString(clusterPV.UnstructuredContent(), "spec.claimRef.name")
+		if err != nil {
+			return false, errors.Wrapf(err, "error looking up persistentvolumeclaim for in-cluster PV %s", name)
+		}
+
+		// Have to create the PVC client here because we don't know what namespace we're using til we get to this point.
+		pvcClient := ctx.coreClient.PersistentVolumeClaims(namespace)
```

Comment:
Testing this right now would require either generating my own fake CoreV1 and PersistentVolumeClaims clients, or bringing in all of client-go's fakes. I don't really want to do either of those, though writing our own seems lighter weight.

Would using `ctx.DynamicFactory` work in it's place? This was originally changed to a typed client since we know exactly what we want, but maybe a dynamic client would be more practical from the testing perspective.

Reply:
Correct, just the deletion timestamp there. Deletion timestamp and phase for namespaces.

Reply:
I think the dynamic client is probably fine for those.